packages/server/socketHandlers/handleMessage.ts

```diff
@@ -29,19 +30,45 @@ const handleParsedMessage = async (
     const response = await handleGraphQLTrebuchetRequest(msg, connectionContext)
     if (response) {
       const {type, id: opId, payload} = response
-      sendGQLMessage(socket, type, payload, opId)
+      sendGQLMessage(connectionContext, type, false, payload, opId)
     }
   })
 }
 
 const PONG = 65
+const ACK = 0
+const REQ = 1
+const MASK = 1
 const isPong = (message) => message.byteLength === 1 && Buffer.from(message)[0] === PONG
-const handleMessage = (websocket: WebSocket, message: ArrayBuffer) => {
+const isAck = (message) =>
+  message.byteLength === 4 && (Buffer.from(message).readUInt32LE() & MASK) === ACK
+const isReq = (message) =>
+  message.byteLength === 4 && (Buffer.from(message).readUInt32LE() & MASK) === REQ
+
+const handleMessage = (
+  websocket: WebSocket | {connectionContext: ConnectionContext},
+  message: ArrayBuffer
+) => {
   const {connectionContext} = websocket
   if (isPong(message)) {
     keepAlive(connectionContext)
     return
   }
+  if (isAck(message)) {
+    const mid = Buffer.from(message).readUInt32LE() >> 1
+    const timer = connectionContext.reliableQueue[mid].timer
```

Comment:
never trust the client!

maybe i'm a sneaky & i tell you the `mid` is 2198759
reliableQueue[mid] will be undefined & your server will crash.